distributed/protocol/utils.py

```diff
@@ -67,6 +67,9 @@ def merge_frames(header, frames):
     if all(len(f) == l for f, l in zip(frames, lengths)):
         return frames
 
+    if len(lengths) == 1 and len(frames) == 1:
```

Comment:
I think I may have been stupid here: is there any way that this condition can be triggered, given the assert and the previous if? If there is only one frame and one length, that length must be the same as the length of the one frame.

If this block came before the previous, then we would avoid calculating lengths again - which should be instantaneous in any case, so should we just get rid of it?

(indeed, the `if all...` condition also catches `if not frames` above, and would be fast to iterate over a zero-length list; and the assert is only needed after that if...)

Reply:
Actually I think you were right to request this and it can come up. For frames that are not too large and for situations where only a single frame is needed, this could be quite common. For example NumPy arrays would pass through this path. So having a fast path makes sense. This saves us going through the `while`, which does a fair bit of work.

Well we would hope the frame is the same size as the length specified. However that might not be true if the message was truncated or otherwise corrupted (and I think this is why the `assert`s are there). So having the `assert`s makes sense. If we moved this before the `assert`s, I think we should add a check to ensure the frame is the expected length.

I might not be following this last bit, but I think it could make sense to move the `if not frames` after the `assert`s or drop it all together. No strong thoughts here. Happy to leave it.

Reply:
Ah I see what you mean (I think). [This check]( https://github.com/dask/distributed/blob/a71951bc60f13054d50af4ff479fd1fbc64b13ad/distributed/protocol/utils.py#L67-L68 ) already acts as a fast path. Yeah we can drop our case.

Reply:
Ok pushed two commits that drop these.